airbyte-integrations/connectors/source-talkdesk-explore/source_talkdesk_explore/streams.py

```diff
@@ -0,0 +1,241 @@
+import requests
+from datetime import datetime
+from abc import abstractmethod
+
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.logger import AirbyteLogger
+
+
+logger = AirbyteLogger()
+
+class GenerateReportStream(HttpStream):
+    """This stream is specifically for generating the report in Talkdesk.
+    - HTTP method: POST
+    - Returns: ID of the generated report
+
+    """
+    primary_key = None
+
+    def __init__(self, base_path, start_date, timezone, **kwargs):
+        super().__init__(**kwargs)
+        self.base_path = base_path
+        self.start_date = start_date
+        self.timezone = timezone
+
+    @property
+    def url_base(self) -> str:
+        return f"https://api.talkdeskapp.com/data/"
+    
+    @property
+    def http_method(self) -> str:
+        return "POST"
+    
+    def path(self, **kwargs) -> str:
+        return self.base_path
+    
+    def request_body_json(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Optional[Mapping]:
+        now = datetime.now().strftime("%Y-%m-%dT%H:%M:%S")
+        logger.info(f"Generating {self.base_path} report from '{self.start_date}' to '{now}'")
+        return {
+            "format": "json",
+            "timespan": {
+                "from": self.start_date,
+                "to": now,
+                "timezone": self.timezone,
+            }
+        }
+    
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+    
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        id_obj = {"id": response_json['job']['id']}
+        logger.info(f"Generated report with ID '{id_obj['id']}'")
+
+        return [id_obj]
+
+
+class ReadReportStream(HttpStream):
+    primary_key = None
+
+    def __init__(self, start_date, timezone, **kwargs):
+        super().__init__(**kwargs)
+        self.start_date = start_date
+        self.timezone = timezone
+
+    @property
+    def url_base(self) -> str:
+        return f"https://api.talkdeskapp.com/data/"
+    
+    def path(self, **kwargs) -> str:
+        latest_state = kwargs.get('stream_state').get(self.cursor_field, None)
+
+        if not latest_state:
+            latest_state = self.start_date
+        
+        # Check and set latest_state to necessary date-time format
+        try:
+            datetime.strptime(latest_state, "%Y-%m-%dT%H:%M:%S")
+        except ValueError:
+            try:
+                datetime.strptime(latest_state, "%Y-%m-%d %H:%M:%S")
+                latest_state = latest_state.replace(" ", "T")
+            except ValueError:
+                logger.error("stream_state is in unhandled date-time format. Required format: %Y-%m-%dT%H:%M:%S")
```

Comment:
Small suggestion @alex-magno WDYT?
```python
if not latest_state:
    # The TalkDesk sometimes return a date-time without the correct format
    latest_state = self.start_date.replace(" ", "T")
if not re.match(latest_state, "%Y-%m-%dT%H:%M:%S"):
    raise ValueError("stream_state doesn't have the right date-time format.")

Reply:
The `latest_state` would usually come from the stream state itself:
```python
latest_state = kwargs.get('stream_state').get(self.cursor_field, None)
```
The only case when it is not set is on the first run - in that case the stream will use the `start_date` instead. And the `start_date` is always validated because it is a parameter of `spec.json`. So we dont need to correct the `start_date`.

The thing with Talkdesk API is that for some streams, the `cursor_field` will be in the format `%Y-%m-%dT%H:%M:%S` but for other streams the format will be `%Y-%m-%d %H:%M:%S`. And only the first format is accepted in the POST request to generate the report. So if we receive the second format, it needs to be corrected.

And about using regex match for that validation: I usually trust `datetime` more when validating dates because it accounts for impossible dates as well (like 2020-01-45, for instance). And I dont think that `%Y-%m-%dT%H:%M:%S` is a regex pattern that is going to work to validate dates. For that you would need something like: `^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})$`. But again, that would not account for impossible dates, only for the amount of digits 😆